src/tools/miri/src/machine.rs

```diff
@@ -552,6 +562,15 @@ pub struct MiriMachine<'mir, 'tcx> {
     /// The spans we will use to report where an allocation was created and deallocated in
     /// diagnostics.
     pub(crate) allocation_spans: RefCell<FxHashMap<AllocId, (Span, Option<Span>)>>,
+
+    /// We do not want to guarantee that the address of a [`mir::Const`] is the same every time it
+    /// is observed, but selecting a new [`AllocId`] (and thus base address) every time the
+    /// [`mir::Const`] is evaluated results in enormous memory use as we must track an unbounded
+    /// number of addresses in a program that repeatedly evaluates one const.
+    /// To avoid this, we compute a handful of [`AllocId`]s for each pair of [`mir::Const`] and the
+    /// [`Instance`] it was evaluated in, then pick between those based on
+    /// [`FrameExtra::frame_id`].
+    const_cache: RefCell<FxHashMap<(mir::Const<'tcx>, Instance<'tcx>), Vec<AllocId>>>,
```

Comment:
What I thought this would be is just `FxHashMap<(mir::Const, usize), AllocId>`, where the `usize` is the frame salt. Why are there `Vec` here, and why is there an `Instance`?